### ISSUE TITLE
Combining protocols and TypedDict

### ISSUE THREAD

**Issue by dsuch** [microsoft/pyright#2616]
Hello,

could you please help me understand the underlying reason for the error message below?

My intention is to be able to express the idea that both certain business objects and dicts - the latter statically defined using TypedDict - conform to a specific protocol.

In this particular case the protocol consists of a series of attributes rather than methods but this is still within the boundaries of [PEP 544](https://www.python.org/dev/peps/pep-0544/) that allows for protocol members to be either.

This is Python 3.8.10 and pyright 1.1.190 running under default settings.

```python

# stdlib
from typing import Protocol, TypedDict

class MessageProto(Protocol):
    msg_id: str

class Message:
    msg_id: str

class DictMessage(TypedDict):
    msg_id: str

# Works nicely
msg1:MessageProto = Message()
msg1.msg_id = 'Hello'

# This is fine as well
msg2:DictMessage = DictMessage(msg_id='World')

# This will return an error
msg3:MessageProto = DictMessage(msg_id='Moon')
```

Here is the error information:

```
$ pyright proto1.py
[...]
/tmp/proto1.py
  /tmp/proto1.py:21:21 - error: Expression of type "DictMessage" cannot be assigned to declared type "MessageProto"
    "DictMessage" is incompatible with protocol "MessageProto"
      "msg_id" is not present (reportGeneralTypeIssues)
1 error, 0 warnings, 0 infos
```

Thank you.

**Comment by dsuch** [comment]
Right, I think I understand it.

The definition using TypedDict says that such a key will be present. On the other hand, the protocol says that an attribute of that name should exist.

Since the dict has no such attribute, the error is returned. Do I interpret it correctly?

All of that is happening within the context of already existing code where the actual runtime recipient of data does not differentiate between attribute vs. key-based access, employing some ``__getitem__`` and ``__getattribute__`` logic to handle both, hence my attempt to specify a protocol with a TypedDict for that.

**Comment by erictraut** [comment]
Yes, your interpretation is correct.
